### PR TITLE
Asynchronous-backing compatible Aura, not plugged in

### PR DESCRIPTION
**Pull request by rphmeier**
Closes https://github.com/paritytech/cumulus/issues/841

This PR prepares the logic for Aura for asynchronous backing according to #2476 but we will still require some updates to how the Polkadot node receives collations in order to fully activate.

The implementation listens for new relay-chain blocks and finds the best parachain block currently descending from the included block as-of that chain. It also selects the slot and timestamp according to the relay-parent's BABE slot (and modifies the relay-driven variant to do the same).

While I originally thought that we'd introduce new block announcement logic, I've decided to go with a different approach:
  * Accept all block announcements
  * Perform equivocation detection / sybil-resistance at the block import stage

The main reasons for this are that equivocation detection isn't best done at the block announcement stage, nor is it safe to deny blocks just because they are equivocations (other authors might receive those equivocations in a different order and build on top of them). This is also more congruent to what is done in Substrate Aura/BABE.

The bulk of the diff comes from refactoring - lots of functionality is common to the async-backing logic and the backwards-compatible collation logic.

TODO:
  - [x] Logic for recursively searching for the parachain-parent
  - [x] Submit collations to the Polkadot side (deferred to after https://github.com/paritytech/polkadot/issues/5056)
  - [x] Rename block announce validator to make it clear it's optional
  - [x] Introduce a new block announce validator which ignores old block announcement data
  - [x] Introduce new runtime API for unincluded segment + Aura
  - [x] Basic sybil/equivocation resistance

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2830529

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2830567

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2839658

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2875432

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2926198

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2956423

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2956965

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2957020

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/3027724

File: client/consensus/aura/src/collators/basic.rs [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+		);
+
+		let (collation, _, post_hash) = try_request!(
+			collator
+				.collate(
+					&parent_header,
+					&claim,
+					None,
+					(parachain_inherent_data, other_inherent_data),
+					// TODO [https://github.com/paritytech/cumulus/issues/2439]
```
Comment (BradleyOlson64) [inline client/consensus/aura/src/collators/basic.rs:183]: Perhaps we can get rid of this TODO. Seems that you handled authoring_duration in the collator params of lookahead.rs

File: client/network/src/lib.rs [hunk 2]
```diff
@@ -452,3 +461,51 @@ async fn wait_to_announce<Block: BlockT>(
...
 	}
 }
+
+/// A [`BlockAnnounceValidator`] which accepts all block announcements, as it assumes
+/// sybil resistance is handled elsewhere.
+#[derive(Debug, Clone)]
+pub struct AssumeSybilResistance(bool);
+
+impl AssumeSybilResistance {
+	/// Instantiate this block announcement validator while permissively allowing (but ignoring)
```
Comment (BradleyOlson64) [inline client/network/src/lib.rs:471]: The "(but ignoring)" from this line seems like it clashes a bit with "shouldn't be ignored" from line 475. 

Perhaps the difference is at which step we ignore certain announcements?

**Review by BradleyOlson64 — APPROVED** [review]
👍  Only nits and questions from me.

**Comment by BradleyOlson64** [comment]
@seunlanlege Implements a slightly more general version of `relay-chain-interface::header()` [here](https://github.com/paritytech/cumulus/pull/2794). 

Perhaps his could be incorporated.

File: client/consensus/aura/src/equivocation_import_queue.rs [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+					}
+				}
+			}
+		}
+
+		Ok(block_params)
+	}
+}
+
+fn slot_now(slot_duration: SlotDuration) -> Slot {
```
[the remark this answers is not in the document]
Comment (rphmeier) [reply client/consensus/aura/src/equivocation_import_queue.rs:207]: AFAIK they have the exact same logic.

File: client/consensus/common/src/lib.rs [hunk 4]
```diff
@@ -181,3 +189,193 @@ where
...
+		let is_pending =
+			entry.depth == 1 && pending_hash.as_ref().map_or(false, |h| &entry.hash == h);
+		let is_included = entry.depth == 0;
+
+		// note: even if the pending block or included block have a relay parent
+		// outside of the expected part of the relay chain, they are always allowed
+		// because they have already been posted on chain.
+		let is_potential = is_pending || is_included || {
+			let digest = entry.header.digest();
+			cumulus_primitives_core::extract_relay_parent(digest).map_or(false, is_hash_in_ancestry) ||
```
[the remark this answers is not in the document]
Comment (rphmeier) [reply client/consensus/common/src/lib.rs:319]: We actually don't produce relay-parent digests in any version of Cumulus, as we have no way of reliably doing that until the persisted validation data and execution parameters are updated to include the relay-parent hash. So for now, every runtime will only produce the relay-parent state root digests. This code is future-proofed to when we can produce those.

File: client/network/src/lib.rs (hunk 2 above)
Comment (rphmeier) [reply client/network/src/lib.rs:471]: Yeah, that's right. Better phrased below would be "shouldn't be rejected"

File: client/network/src/lib.rs [hunk 5]
```diff
@@ -452,3 +461,51 @@ async fn wait_to_announce<Block: BlockT>(
...
+	/// announcements which come tagged with seconded messages.
+	///
+	/// This is useful for backwards compatibility when upgrading nodes: old nodes will continue
+	/// to broadcast announcements with seconded messages, so these announcements shouldn't be ignored
```
Comment (rphmeier) [inline client/network/src/lib.rs:475]:
```suggestion
	/// to broadcast announcements with seconded messages, so these announcements shouldn't be rejected
```

File: client/consensus/aura/src/collators/basic.rs (hunk 1 above)
Comment (rphmeier) [reply client/consensus/aura/src/collators/basic.rs:183]: Unfortunately, that doesn't really work for the 'basic' collator as the time limitations are very strict (and this PR just maintains behavior as this code is becoming obsolete)

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/3129693

File: client/consensus/common/src/lib.rs [hunk 6]
```diff
@@ -181,3 +192,193 @@ where
...
+/// for parachain blocks. The search begins at the last included parachain block and returns
+/// a set of [`PotentialParent`]s which could be potential parents of a new block with this
+/// relay-parent according to the search parameters.
+///
+/// A parachain block is a potential parent if it is either the last included parachain block, the pending
+/// parachain block (when `max_depth` >= 1), or all of the following hold:
+///   * its parent is a potential parent
+///   * its relay-parent is within `ancestry_lookback` of the targeted relay-parent.
+///   * the block number is within `max_depth` blocks of the included block
+pub async fn find_potential_parents<B: BlockT>(
```
Comment (slumber) [inline client/consensus/common/src/lib.rs:240]: Can we at least file a follow up to add tests for this?

**Review by slumber — APPROVED** [review]

File: client/consensus/common/src/lib.rs (hunk 6 above)
Comment (rphmeier) [reply client/consensus/common/src/lib.rs:240]: Yes, https://github.com/paritytech/cumulus/issues/2831

File: client/consensus/aura/src/collators/lookahead.rs [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+		// If the longest chain has space, build upon that. Otherwise, don't
+		// build at all.
+		potential_parents.sort_by_key(|a| a.depth);
+		let initial_parent = match potential_parents.pop() {
+			None => continue,
+			Some(p) => p,
+		};
+
+		// Build in a loop until not allowed. Note that the authorities can change
+		// at any block, so we need to re-claim our slot every time.
```
[the remark this answers is not in the document]
Comment (slumber) [reply client/consensus/aura/src/collators/lookahead.rs:234]: >  Is that what you mean, or something else?

Yes. Concern is this behavior might not be something parachains would expect.

[the remark this answers is not in the document]
Comment (rphmeier) [reply client/consensus/aura/src/collators/lookahead.rs:234]: There is really no other option - either you produce blocks at a constant rate, and never build a backlog, or you build an infinite backlog. We do need to communicate these properties, I agree, but based on conversations and other socialization of the concept that it's pretty uncontroversial.

**Review by slumber — APPROVED** [review]

**Pull request merged**